examples/metrics/metrics/observer.js

```diff
@@ -61,7 +61,7 @@ meter.createBatchObserver((observerBatchResult) => {
     });
   }, {
     maxTimeoutUpdateMS: 500,
-    logger: new ConsoleLogger(LogLevel.DEBUG)
+    logger: diagLogLevelFilter(DiagLogLevel.DEBUG, new DiagConsoleLogger())
```

Comment:
If we follow the same factory model (as with the NoopLogger) this would become (perhaps)
```logger: createLogLevelDiagLogger(DiagLogLevel.DEBUG, new DiagConsoleLogger())```

The DiagnosticAPI does already of a setLogLevel() to enforce a global default logging level. As part of the SDK there is a Global LOG_LEVEL but it hangs off environment so that s a different discussion.

Reply:
Additionally, once we only use diag this will become
```
diag.setLogger(new DiagConsoleLogger());
diag.setLogLevel(DiagLogLevel.DEBUG);
```

And then it would need to be added here.

Reply:
tbh it looks strange that you still have to create a new `DiagConsoleLogger()` which you then have to put to the "factory" to create a new `LogLevelDiagLogger`.  Why do I have to create a 2 instances of 2 different classes to be able to have a logger with desired log level. 

Things like 
`logger: new ConsoleLogger(LogLevel.DEBUG)`
is self explanatory easy and intuitive, why can't we have
one of:
```
logger: createDiagConsoleLogger(DiagLogLevel.DEBUG)
// or
logger: new DiagConsoleLogger(DiagLogLevel.DEBUG)
// or
logger: new DiagConsoleLogger().setLogLevel(DiagLogLevel.DEBUG)

```
comparing to
```
logger: createLogLevelDiagLogger(DiagLogLevel.DEBUG, new DiagConsoleLogger())
```

Reply:
This is part of the design to reduce code duplication, doing it this way NO DiagLogger (except the DiagLogger returned by createLogLevelDiagLogger()) needs to add any code to enforce log level filtering.

So ALL log level filtering is in a single location and anyone who want to create their own filter (file, database, event logger, etc) just doesn't need to worry about it.

The only other solution to achieve the same would be to introduce a Base class that everyone would need to extend and this has a couple of implications, not just on size but the actual function implementation of the Base class would need to be prototype level OR we force every implementor to save / call the base instance level functions.

Reply:
And generally, aren't we also saying that the code moving forward (and unless absolutely necessary) would not create it's own instances and instead would use the ```api.diag``` instance -- this will be a migration process and
```
diag.setLogger(new DiagConsoleLogger());
diag.setLogLevel(DiagLogLevel.DEBUG);
```

Reply:
Personally, I think this distinction makes sense and it is similar to previous loggers i've seen. I think confusion may primarily come from the naming. What I have seen in the past is `LoggingDestination` or `LogSink`. Here's a very simplified example to help explain what is going on and why it is implemented this way.

```
interface LogSink {
    debug(...args): void;

    info(...args): void;

    warn(...args): void;

    error(...args): void;

    critical(...args): void;
}

enum LogLevel {
    DEBUG = 0,
    INFO = 10,
    WARN = 20,
    ERROR = 30,
    CRITICAL = 40,
}

class Logger {
    constructor(
        private level: LogLevel,
        private sink: LogSink,
    ) { }

    debug(...args) {
        if (this.level < LogLevel.DEBUG)
        this.sink.debug(...args)
    }

    info(...args) {
        if (this.level < LogLevel.INFO)
        this.sink.info(...args)
    }

    warn(...args) {
        if (this.level < LogLevel.WARN)
        this.sink.warn(...args)
    }

    error(...args) {
        if (this.level < LogLevel.ERROR)
        this.sink.error(...args)
    }

    critical(...args) {
        if (this.level < LogLevel.CRITICAL)
        this.sink.critical(...args)
    }
}


// console sink does not need to do any filtering. this is already done
class ConsoleLogSink implements LogSink {
    debug(...args) {
        console.log(...args);
    };

    info(...args) {
        console.log(...args);
    };

    warn(...args) {
        console.error(...args);
    };

    error(...args) {
        console.error(...args);
    };

    critical(...args) {
        console.error(...args);
    };
}


// file sink does not need to do any filtering. this is already done
class FileLogSink implements LogSink {
    constructor(private filename: string) { }

    debug(...args) {
        fs.appendFileSync(this.filename, format(...args))
    };

    info(...args) {
        fs.appendFileSync(this.filename, format(...args))
    };

    warn(...args) {
        fs.appendFileSync(this.filename, format(...args))
    };

    error(...args) {
        fs.appendFileSync(this.filename, format(...args))
    };

    critical(...args) {
        fs.appendFileSync(this.filename, format(...args))
    };
}


// User can now use either logging sink/destination without them both having to implement filtering based on log levels
const fileLogger = new Logger(LogLevel.INFO, new FileLogSink("info.log"));
const consoleLogger = new Logger(LogLevel.CRITICAL, new ConsoleLogSink());

```

Reply:
thx the "sink" approach / explanation is clear and straightforward